src/v/pandaproxy/schema_registry/sharded_store.cc

```diff
@@ -0,0 +1,227 @@
+sharded_store::get_compatibility(const subject& sub) {
+    using overload_t = result<compatibility_level> (store::*)(const subject&)
+      const;
+    auto level = co_await _store.invoke_on(
+      shard_for(sub), static_cast<overload_t>(&store::get_compatibility), sub);
+    co_return level.value();
+}
+
+ss::future<bool>
+sharded_store::set_compatibility(compatibility_level compatibility) {
+    auto map = [compatibility](store& s) {
+        return s.set_compatibility(compatibility).value();
+    };
+    auto reduce = std::logical_and<>{};
+    co_return co_await _store.map_reduce0(map, true, reduce);
+}
+
+ss::future<bool> sharded_store::set_compatibility(
+  const subject& sub, compatibility_level compatibility) {
+    using overload_t = result<bool> (store::*)(
+      const subject&, compatibility_level);
+    auto set = co_await _store.invoke_on(
+      shard_for(sub),
+      static_cast<overload_t>(&store::set_compatibility),
+      sub,
+      compatibility);
+    co_return set.value();
+}
+
+ss::future<bool> sharded_store::clear_compatibility(const subject& sub) {
+    auto cleared = co_await _store.invoke_on(
+      shard_for(sub), &store::clear_compatibility, sub);
+    co_return cleared.value();
+}
+
+ss::future<sharded_store::insert_schema_result>
+sharded_store::insert_schema(schema_definition def, schema_type type) {
+    auto map = [&def, type](store& s) { return s.get_schema_id(def, type); };
+    auto reduce = [](
+                    std::optional<schema_id> acc,
+                    std::optional<schema_id> s_id) { return acc ? acc : s_id; };
+    auto s_id = co_await _store.map_reduce0(
+      map, std::optional<schema_id>{}, reduce);
+    if (s_id) {
+        co_return insert_schema_result{*s_id, false};
+    }
+
+    auto new_s_id = co_await allocate_schema_id();
+    auto inserted = co_await upsert_schema(new_s_id, std::move(def), type);
+    co_return insert_schema_result{new_s_id, inserted};
+}
+
+ss::future<bool> sharded_store::upsert_schema(
+  schema_id id, schema_definition def, schema_type type) {
+    co_await maybe_update_max_schema_id(id);
+    co_return co_await _store.invoke_on(
+      shard_for(id), &store::upsert_schema, id, std::move(def), type);
+}
+
+ss::future<sharded_store::insert_subject_result>
+sharded_store::insert_subject(subject sub, schema_id id) {
+    auto [version, inserted] = co_await _store.invoke_on(
+      shard_for(sub), &store::insert_subject, sub, id);
+    co_return insert_subject_result{version, inserted};
+}
+
+ss::future<bool> sharded_store::upsert_subject(
+  subject sub, schema_version version, schema_id id, is_deleted deleted) {
+    co_return co_await _store.invoke_on(
+      shard_for(sub), &store::upsert_subject, sub, version, id, deleted);
+}
+
+ss::future<schema_id> sharded_store::allocate_schema_id() {
+    auto increment = [this] { return _next_schema_id++; };
+    co_return co_await ss::smp::submit_to(
+      ss::shard_id{0}, _smp_opts, std::move(increment));
+}
```

Comment:
this seems ok, but it's also a sign that you might benefit from one or more service abstractions existing at a higher level since it would appear that you have state `_next_schema_id` that is stored on every core but only accessed/modified on a single core.

Reply:
> this seems ok, but it's also a sign that you might benefit from one or more service abstractions existing at a higher level since it would appear that you have state `_next_schema_id` that is stored on every core but only accessed/modified on a single core.

Yeah, that's a little awkward. I need an atomic cross-core increment-only id allocator.

Reply:
yeh its totally fine. if the state were a little bigger you could store it on the heap on core-0, but no benefit here.

Reply:
btw how do you seed this value and how is it used? you could also generate unique ids by using core-id in the high bits and then increment locally on your core.

Reply:
> btw how do you seed this value and how is it used? could you generate unique ids by using core-id in the high bits and then increment locally on your core.

It starts at 1 and increments for each unique schema (roughly a lexicographical comparison of the canonical form of the schema and its type)

I was aiming for some compatibility here, but it's a little tricky to canonicalize the schema (the c++ AVRO library doesn't help much here, but the spec is clear enough).

The public REST API is a little tricky since schema aren't registered directly, and the schema itself gets serialised into each subject-version on the topic (probably for compaction reasons).

There probably are more efficient mechanisms for choosing an unique ID, but since there's only one leader (for now), I'm keeping it simple. I don't expect the mutable endpoints to be particularly hot.

Reply:
yeh that sounds good. last question: does the starting value for the unique id counter have to be seeded from the state restored from a log that is read when the leader starts up?

Reply:
Currently one leader is assumed. The `ensure_started` stuff blocks until the entire topic is read, and from then on it's in-memory, plus write-to-topic if a diff is required. The HA stuff requires some changes here.

Reply:
mm, I think my question is not clear. assuming one leader, assume some schemas are created resulting in unique ids being used up to 100. these schemas are stored in the backing topic. the leader restarts, reads this log, and then needs to start handling new requests. does it need to seed the unique id counter at 101, or does it start back at 1? if the later, is that problematic since now it would not seem to be a unique id?

Reply:
Due to the fact that the topic is read before requests are served, and that applying the messages from the topic to the store is an upsert, and taking into consideration that topics are compacted, and given this function: https://github.com/vectorizedio/redpanda/pull/1742/files#diff-9b182dbde9ea6139bf35354502b76f869ca2ea668ce6c2e1a70ffb40749abef0R219 - the next ID should be sequential, under the current assumption that there is but one leader.

The HA changes will remove the requirement of the assumption, and will use a belt-and-braces approach to leadership transfer through (starting with whether the member of a consumer group has any assignment over the single TP - the assumption that it is the primary), via the use of produce transactions, and in particular the violation of the produce epoch. At least, that's my plan and current understanding of produce transactions.

Reply:
> Due to the fact that the topic is read before requests are served, and that applying the messages from the topic to the store is an upsert, and taking into consideration that topics are compacted, and given this function: https://github.com/vectorizedio/redpanda/pull/1742/files#diff-9b182dbde9ea6139bf35354502b76f869ca2ea668ce6c2e1a70ffb40749abef0R219, under the basis that it's dispatched to a single core, the next ID should be sequential, under the current assumption that there is but one leader.

thanks. the key thing that was missing for me was `maybe_update_max_schema_id` which pushed the id forward, presumably as part of the log replay on startup.